mist/runtime/loss.py

```diff
@@ -31,6 +31,64 @@ def forward(self, y_true, y_pred):
         loss = torch.mean(loss)
         return loss
 
+#Implement clDice:
+class soft_cldice(nn.Module):
+    def __init__(self, iter_=3, smooth = 1., exclude_background=False):
+        super(soft_cldice, self).__init__()
+        self.iter = iter_
+        self.smooth = smooth
+        self.soft_skeletonize = SoftSkeletonize(num_iter=10)
+        self.exclude_background = exclude_background
+
+    def forward(self, y_true, y_pred):
+        if self.exclude_background:
+            y_true = y_true[:, 1:, :, :]
+            y_pred = y_pred[:, 1:, :, :]
+        skel_pred = self.soft_skeletonize(y_pred)
+        skel_true = self.soft_skeletonize(y_true)
+        tprec = (torch.sum(torch.multiply(skel_pred, y_true))+self.smooth)/(torch.sum(skel_pred)+self.smooth)    
+        tsens = (torch.sum(torch.multiply(skel_true, y_pred))+self.smooth)/(torch.sum(skel_true)+self.smooth)    
+        cl_dice = 1.- 2.0*(tprec*tsens)/(tprec+tsens)
+        return cl_dice
+
+
+def soft_dice(y_true, y_pred):
+    """[function to compute dice loss]
+
+    Args:
+        y_true ([float32]): [ground truth image]
+        y_pred ([float32]): [predicted image]
+
+    Returns:
+        [float32]: [loss value]
+    """
+    smooth = 1
+    intersection = torch.sum((y_true * y_pred))
+    coeff = (2. *  intersection + smooth) / (torch.sum(y_true) + torch.sum(y_pred) + smooth)
+    return (1. - coeff)
+
+
+class soft_dice_cldice(nn.Module):
+    def __init__(self, iter_=3, alpha=0.5, smooth = 1., exclude_background=False):
+        super(soft_dice_cldice, self).__init__()
+        self.iter = iter_
+        self.smooth = smooth
+        self.alpha = alpha
+        self.soft_skeletonize = SoftSkeletonize(num_iter=10)
+        self.exclude_background = exclude_background
+
+    def forward(self, y_true, y_pred):
+        if self.exclude_background:
```

Comment:
Also, add these lines here:

```
# Prepare inputs
y_true = get_one_hot(y_true, y_pred.shape[1])
y_pred = softmax(y_pred, dim=1)
```

We apply `get_one_hot` because the data loader does not stream the masks in one hot encoder and the networks don't apply a final activation to their outputs. This latter part gives us more flexibility for future applications like regression instead of segmentation.

Reply:
Is this using the same implementation as the ```SoftCLDice```? If so, then create an instance of ```SoftCLDice``` in ```__init``` and call it in the forward pass of this loss. Something like this should do the same thing right? It's always a best practice to reuse your existing implementations if you can. In this case, try reusing your ```SoftCLDice```. Other than this, everything else looks great and we should be ready to merge!

```
class SoftDiceCLDice(nn.Module):
        def __init__(self, iterations=3, alpha=0.5, smooth=1., exclude_background=False):
            super(SoftDiceCLDice, self).__init__()
            self.iterations = iterations
            self.smooth = smooth
            self.alpha = alpha
            self.dice_loss = DiceLoss()
            self.exclude_background = exclude_background
            self.cldice_loss = SoftCLDice(self.iterations, self.smooth, self.exclude_background)

       def forward(self, y_true, y_pred):
           dice = self.dice_loss(y_true, y_pred)
           cldice = self.cldice_loss(y_true, y_pred)
           return (1.0 - self.alpha) * dice + self.alpha * cl_dice
```